server/src/main/java/org/elasticsearch/index/query/VectorGeoPointShapeQueryProcessor.java

```diff
@@ -0,0 +1,181 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.document.LatLonPoint;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.geo.GeoPolygonDecomposer;
+import org.elasticsearch.common.geo.GeoShapeType;
+import org.elasticsearch.common.geo.ShapeRelation;
+import org.elasticsearch.geometry.Circle;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.GeometryCollection;
+import org.elasticsearch.geometry.GeometryVisitor;
+import org.elasticsearch.geometry.LinearRing;
+import org.elasticsearch.geometry.MultiLine;
+import org.elasticsearch.geometry.MultiPoint;
+import org.elasticsearch.geometry.MultiPolygon;
+import org.elasticsearch.geometry.Point;
+import org.elasticsearch.geometry.Polygon;
+import org.elasticsearch.geometry.Rectangle;
+import org.elasticsearch.geometry.ShapeType;
+import org.elasticsearch.index.mapper.AbstractGeometryFieldMapper;
+import org.elasticsearch.index.mapper.GeoPointFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+
+import java.util.ArrayList;
+
+import static org.elasticsearch.index.mapper.GeoShapeIndexer.toLucenePolygon;
+
+public class VectorGeoPointShapeQueryProcessor implements AbstractGeometryFieldMapper.QueryProcessor {
+
+    @Override
+    public Query process(Geometry shape, String fieldName, ShapeRelation relation, QueryShardContext context) {
+        validateIsGeoPointFieldType(fieldName, context);
+        // geo points only support intersects
+        if (relation != ShapeRelation.INTERSECTS) {
+            throw new QueryShardException(context,
+                relation+ " query relation not supported for Field [" + fieldName + "].");
+        }
+        // wrap geoQuery as a ConstantScoreQuery
+        return getVectorQueryFromShape(shape, fieldName, relation, context);
+    }
+
+    private void validateIsGeoPointFieldType(String fieldName, QueryShardContext context) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        if (fieldType instanceof GeoPointFieldMapper.GeoPointFieldType == false) {
+            throw new QueryShardException(context, "Expected " + GeoPointFieldMapper.CONTENT_TYPE
+                + " field type for Field [" + fieldName + "] but found " + fieldType.typeName());
+        }
+    }
+
+    protected Query getVectorQueryFromShape(
+        Geometry queryShape, String fieldName, ShapeRelation relation, QueryShardContext context) {
+        ShapeVisitor shapeVisitor = new ShapeVisitor(context, fieldName, relation);
+        if (queryShape.type().equals(ShapeType.POLYGON)) {
+            ArrayList<org.elasticsearch.geometry.Polygon> collector = new ArrayList<>();
```

Comment:
I think we should move this logic inside the visitor and decompose the polygons in the methods `visit(MultiPolygon multiPolygon)` and `visit(Polygon polygon)`. Note that the query accepts an array and therefore the boolean query is not necessary.

Reply:
I removed GeoPolygonDecomposer.decomposePolygon() prior to pushing this down & tests passed OK where I had expected a fail.  I added an additional point to GeoShapeQueryTests.testPolygonSpanningDateline() to validates that a polygon spanning the dateline matched intersecting documents on both sides and did not match non-intersecting documents on both sides - the test is currently passing without decomposePolygon() , I'm looking into how and why

Reply:
I guess it is because the method lives in `GeoShapeQueryTests` so it is only testing `geo_shape` field. Try to move it down to `geoQueryTests` (you might need to change the way you are indexing the points)

Reply:
Tests were in GeoShape, I've pushed up, working on the refactor now

Reply:
Done - by the way, in VectorGeoPointShapeQueryProcessor.ShapeVisitor the methods

visit(MultiPolygon multiPolygon)
visit(Polygon polygon)

are both using (successfully) 

org.apache.lucene.geo.Polygon

This is the right level to introduce the Lucene classes, correct?